lib/ts/string_view.h

```diff
@@ -0,0 +1,1199 @@
+  CONSTEXPR14 size_type
+  copy(_Type *const rhs, size_type length, const size_type offset = 0) const
+  {
+    check_offset_bound(offset);
+    length = std::min(length, m_size - offset);
+    traits_type::copy(rhs, m_data + offset, length);
+    return length;
+  }
+
+  CONSTEXPR14 basic_string_view
+  substr(const size_type offset = 0, size_type length = npos) const
+  {
+    check_offset_bound(offset);
+    length = std::min(length, m_size - offset);
+    return basic_string_view(m_data + offset, length);
+  }
+
+  // this is not a std function.. but is needed in some form for ==, != operators
+  // I could make this _equal to say don't call it.. but I won't
+  constexpr bool
+  equal(const basic_string_view rhs) const noexcept
+  {
+    return (m_size == rhs.m_size && traits_type::compare(m_data, rhs.m_data, m_size) == 0);
+  }
+
+  //////////////////////////////////////////////
+  // Compare
+
+  CONSTEXPR14 int
+  compare(const basic_string_view rhs) const noexcept
+  {
+    const int result = traits_type::compare(m_data, rhs.m_data, std::min(m_size, rhs.size()));
+
+    if (result != 0) {
+      return result;
+    }
+
+    if (m_size < rhs.m_size) {
+      return -1;
+    }
+
+    if (m_size > rhs.m_size) {
+      return 1;
+    }
+
+    return 0;
+  }
+
+  constexpr int
+  compare(const size_type offset, const size_type length, const basic_string_view rhs) const
+  {
+    return (substr(offset, length).compare(rhs));
+  }
+
+  constexpr int
+  compare(const size_type offset, const size_type length, const basic_string_view rhs, const size_type rhsoffsetset,
+          const size_type rhs_length) const
+  {
+    return (substr(offset, length).compare(rhs.substr(rhsoffsetset, rhs_length)));
+  }
+
+  constexpr int
+  compare(const _Type *const rhs) const
+  {
+    return (compare(basic_string_view(rhs)));
+  }
+
+  constexpr int
+  compare(const size_type offset, const size_type length, const _Type *const rhs) const
+  {
+    return (substr(offset, length).compare(basic_string_view(rhs)));
+  }
+
+  constexpr int
+  compare(const size_type offset, const size_type length, const _Type *const rhs, const size_type rhs_length) const
+  {
+    return (substr(offset, length).compare(basic_string_view(rhs, rhs_length)));
+  }
+
+  //////////////////////
+  // find -- string
+
+  CONSTEXPR14 size_type find(const _Type *const rhs, const size_type offset, const size_type rhs_length) const
+    noexcept // strengthened
+  {
+    // do we have space to find the string given the offset
+    if (rhs_length > m_size || offset > m_size - rhs_length) {
+      // no match found
+      return npos;
+    }
+
+    if (rhs_length == 0) { // empty string always matches to offset
+      return offset;
+    }
+
+    // this is the point in which we can stop checking
+    const auto end_ptr = m_data + (m_size - rhs_length) + 1;
+    // start looking
+    for (auto curr = m_data + offset;; ++curr) {
```

Comment:
Is there a problem if `end_ptr == curr` for a successful `trait_type::find` but it's not a `compare` match, then `curr` is incremented and the `find` called again with `curr > end_ptr`?